src/terms/nonlocal.jl

```diff
@@ -149,29 +148,28 @@ where pihat(q) = ∫_R^3 pi(r) e^{-iqr} dr
 We store 1/√Ω pihat(k+G) in proj_vectors.
 """
 function build_projection_vectors_(basis::PlaneWaveBasis{T}, atoms, kpt::Kpoint) where {T}
+    unit_cell_volume = basis.model.unit_cell_volume
     n_proj = count_n_proj(atoms)
-    model = basis.model
-
-    proj_vectors = zeros(Complex{T}, length(G_vectors(kpt)), n_proj)
-    qs = [model.recip_lattice * (kpt.coordinate + G) for G in G_vectors(kpt)]
+    n_G    = length(G_vectors(basis, kpt))
+    proj_vectors = zeros(Complex{T}, n_G, n_proj)
 
     # Compute the columns of proj_vectors = 1/√Ω pihat(k+G)
     # Since the pi are translates of each others, pihat(k+G) decouples as
     # pihat(q) = ∫ p(r-R) e^{-iqr} dr = e^{-iqR} phat(q).
     # The first term is the structure factor, the second the form factor.
-    offset = 0 # offset into proj_vectors
+    offset = 0  # offset into proj_vectors
     for (psp, positions) in atoms
         # Compute position-independent form factors
-        form_factors = build_form_factors(psp, qs)
+        form_factors = build_form_factors(psp, Gplusk_vectors_cart(basis, kpt))
 
         # Combine with structure factors
         for r in positions
             # k+G in this formula can also be G, this only changes an unimportant phase factor
-            structure_factors = [cis(-2T(π)*dot(kpt.coordinate + G, r)) for G in G_vectors(kpt)]
-            for iproj = 1:count_n_proj(psp)
-                @views proj_vectors[:, offset+iproj] .= (structure_factors
-                                                        .* form_factors[:, iproj]
-                                                        ./ sqrt(model.unit_cell_volume))
+            structure_factors = map(q -> cis(-2T(π) * dot(q, r)), Gplusk_vectors(basis, kpt))
```

Comment:
Do we "officially" want to use `map` instead of comprehensions? I quite like comprehensions, I find it's more readable than map, but I understand it makes more sense (eg for GPU)...

Reply:
Yeah I think we should at least slowly move towards it, even though I agree comprehensions are generally more readable.

Reply:
the main problem I have with the map is that it's less easy to see that q is in Gplusk

Reply:
we could plausibly have a `@keep_types` macro that rewrites the type of the comprehension or something

Reply:
Hmm. I don't think more macros will make it easier for AD suites etc. to process the code. Also generally comprehensions are type stable and can be type-annotated, so I don't see what such a macro would do differently.